[Volume] RunPod network volume support

DanielZhangQD opened the pull request:
<!-- Describe the changes in this PR -->
Fix https://github.com/skypilot-org/skypilot/issues/6953, Fix #6416
RunPod network volume support

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

<!-- CI commands (/-prefixed) can only be triggered by repo members -->

DanielZhangQD commented:
/smoke-test --kubernetes -k test_volumes_on_kubernetes

DanielZhangQD commented:
/smoke-test --kubernetes -k test_volumes_on_kubernetes

DanielZhangQD commented:
/smoke-test --kubernetes -k test_volumes_on_kubernetes

aylei reviewed (commented):
Did not look into the test code, the application change LGTM overall, thanks @DanielZhangQD !

Michaelvll commented:
Can we split the volume SDK fixes into a separate PR? It seems the volume SDK is for a general k8s volume. The RunPod volume support is a separate feature to add?

DanielZhangQD commented on sky/provision/runpod/instance.py:
```diff
@@ -80,6 +80,21 @@ def run_instances(region: str, cluster_name_on_cloud: str,
+    network_volume_id = None
+    volume_mount_path = None
+    if volume_mounts:
+        if len(volume_mounts) > 1:
+            logger.warning(
+                f'RunPod only supports one network volume mount, '
+                f'but {len(volume_mounts)} are specified. Only the first one '
+                f'will be used.')
```
The check in the `sky/clouds/runpod.py` is a precheck before provisioning, which will ensure only 1 item for now, here is just in case.

DanielZhangQD commented on sky/provision/runpod/volume.py:
```diff
@@ -0,0 +1,225 @@
+            'name': name_on_cloud,
+            'size': size_int,
+        }
+        resp = _rest_request('POST', '/networkvolumes', json=payload)
```
No.

DanielZhangQD commented on sky/client/cli/command.py:
```diff
@@ -4259,6 +4258,24 @@ def volumes_apply(
+def _parse_volume_override_config(
+        name: Optional[str],
+        infra: Optional[str],
+        type: Optional[str],  # pylint: disable=redefined-builtin
```
Updated

DanielZhangQD commented on sky/provision/runpod/volume.py:
```diff
@@ -0,0 +1,225 @@
+        raise RuntimeError(
+            'RunPod API key is not set. Please set runpod.api_key '
+            'or RUNPOD_API_KEY.')
+    return str(api_key)
```
Updated

DanielZhangQD commented on sky/provision/runpod/volume.py:
```diff
@@ -0,0 +1,225 @@
+    return str(api_key)
+
+
+def _rest_request(method: str,
```
Updated

DanielZhangQD commented:
> Can we split the volume SDK fixes into a separate PR? It seems the volume SDK is for a general k8s volume. The RunPod volume support is a separate feature to add?

The volume SDK applies to both RunPod volume and k8s volume. I have added examples for both cases in the comment. In addition, the RunPod volume support relies on this code improvement.

DanielZhangQD commented:
/smoke-test --kubernetes -k test_volumes_on_kubernetes

DanielZhangQD commented:
/quicktest-core

aylei commented:
/quicktest-core --base-branch v0.10.2

kevinmingtarja commented:
/quicktest-core --base-branch v0.10.1 --kubernetes
